### DISCUSSION TITLE
Secure Cookies

### DISCUSSION THREAD

**Discussion by TylerAHolden** [vercel/vercel#6052] [Help]
```
import { NextApiRequest, NextApiResponse } from 'next';
import Cookies from 'cookies';

export default async (req: NextApiRequest, res: NextApiResponse) => {
  try {
    const cookies = new Cookies(req, res);

    cookies.set('__session', 'cookietest', {
      httpOnly: true,
      secure: true,
      sameSite: 'strict',
      domain: 'mydomain.com',
    });

    res.status(200).json({ success: true });
  } catch (err) {
    res.status(500).json({ msg: err.message });
  }
};

```

I try to set a secure cookie via an API serverless route I receive an error: "Cannot send secure cookie over unencrypted connection". I'm using a custom domain but all of the certs are valid so it is going through HTTPS. Is this an issue with something related to Vercel's serverless infrastructure?

**Comment by TylerAHolden** [comment]
Nevermind I found a very helpful article here that explained a different approach: https://maxschmitt.me/posts/next-js-http-only-cookie-auth-tokens/

**Reply by TylerAHolden** [reply]
I cant figure out how to delete this discussion post

**Comment by danechitoaie** [comment]
I'm having the same issue. trying to set a cookie with `secure: true` fails with same error: "Cannot send secure cookie over unencrypted connection".

Is Vercel having some kind of HTTP proxy that's HTTP and that's causing this?

**Reply by danechitoaie** [reply]
If anyone still has this issue see https://github.com/pillarjs/cookies/issues/142

**Reply by sagarnanera** [reply]
Hey @danechitoaie,
Seems like the issue you mentioned above has been deleted.
And i am facing the same issue.

I have deployed my back-end on render and front-end in vercel. I am using koaJS in backend and react in frontend.
So Can you guild me to resolve the cookie issue ?

Thank you.

**Reply by danechitoaie** [reply]
See the message bellow from @dvakatsiienko https://github.com/vercel/vercel/discussions/6052#discussioncomment-5682727

**Comment by dvakatsiienko** [comment]
The solution [pillarjs/cookies/issues/142](https://github.com/pillarjs/cookies/issues/142) worked for me!

The `secure` field needs to be configured both in the Cookies constructor and the instance.

```ts
const cookies = new Cookies(req, res, { secure: process.env.NODE_ENV === 'production' });

cookies.set(NEXT_AUTH_SESSION_TOKEN_COOKIE_NAME, sessionToken, {
    expires: sessionExpiry,
    secure:  process.env.NODE_ENV === 'production',
});
```
